### PR TITLE
Table of contents block accessibility improvements

### PR DESCRIPTION
**Pull request by alexstine**
## What?

This PR fixes some accessibility issues in the table of contents block, front-end and back-end.

## Why?

Accessibility is important.

## How?

- Remove inert.
- Disable the ability to click links in the back-end block.
- Add `aria-label` to front-end `<nav>` tag.

## Testing Instructions

1. Insert a Table of Contents block.
2. Verify the block is keyboard accessible with arrow keys.
3. Verify clicking the links is disabled.
4. Verify that screen readers get a disabled announcement through `aria-disabled="true"` attribute.
5. Check the front end for the `aria-label="Table of Contents"` attribute.

### Testing Instructions for Keyboard

Verify in the back-end that Up and Down Arrow keys allow you to navigate the block content.

### CODE REVIEW

File: packages/block-library/src/table-of-contents/edit.js
Comment (alexstine) [inline packages/block-library/src/table-of-contents/edit.js:140]: I'm not sure why no one caught this regression. This makes the content completely inaccessible to keyboard users. Please don't do this unless you have a really good reason to.

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/inert

File: packages/block-library/src/table-of-contents/edit.js
Comment (alexstine) [inline packages/block-library/src/table-of-contents/edit.js:163]: Passing this prop to disable link activation for the editor side.

File: packages/block-library/src/table-of-contents/list.tsx
Comment (alexstine) [inline packages/block-library/src/table-of-contents/list.tsx:24]: Is there a better way to type this?

File: packages/block-library/src/table-of-contents/list.tsx
Comment (alexstine) [inline packages/block-library/src/table-of-contents/list.tsx:36]: This communicates disabled state to screen readers.

File: packages/block-library/src/table-of-contents/list.tsx
Comment (alexstine) [inline packages/block-library/src/table-of-contents/list.tsx:37]: This prevents click.

File: packages/block-library/src/table-of-contents/list.tsx
Comment (alexstine) [inline packages/block-library/src/table-of-contents/list.tsx]: This prevents context menu open in new tab/window.

File: packages/block-library/src/table-of-contents/save.js
Comment (alexstine) [inline packages/block-library/src/table-of-contents/save.js]: This allows us to pass an accessible name to the front-end.

**Review by alexstine — COMMENTED** [review]
Implementation notes below.

File: packages/block-library/src/table-of-contents/edit.js
Comment (joedolson) [reply packages/block-library/src/table-of-contents/edit.js:140]: `inert` was added in several places in https://github.com/WordPress/gutenberg/commit/d50e613d110ea7ccd5de3f647c8dcf3b72bdda74; we maybe should take a look at each usage.

File: packages/block-library/src/table-of-contents/save.js
Comment (Mamaduka) [reply packages/block-library/src/table-of-contents/save.js]: We can't use translatable strings in the `save` method. It breaks the "purity" of a function, as it can return different values when the same arguments are passed based on user locale. This breaks block validation.

P.S. I'm in the process of significant refactoring of a ToC block, which will unblock this issue. Ref: #54224.

Comment (alexstine) [reply packages/block-library/src/table-of-contents/save.js]: @Mamaduka Had no idea. Why don't we have a static check for that? I can remove the translation, then it should be okay?

Comment (Mamaduka) [reply packages/block-library/src/table-of-contents/save.js]: Not sure, but that’s a good idea.

You can ignore saved content for now. The block will be dynamic after I land my refactoring and I’ll include the label.

Comment (alexstine) [reply packages/block-library/src/table-of-contents/save.js]: @Mamaduka Oh, that's cool. You are building a PHP renderer for the block. Sounds good. I will get rid of this change. 👍

Comment (Mamaduka) [reply packages/block-library/src/table-of-contents/save.js]: Thank you, @alexstine!

Comment (alexstine) [reply packages/block-library/src/table-of-contents/save.js]: Done. 👍

**Comment by Mamaduka** [comment]
The `aria-label` is now included in the server render Table of Contents block - https://github.com/WordPress/gutenberg/pull/54224/commits/b0935ab917c672e046eb82b2e116202c8efe8eaf.

**Comment by Mamaduka** [comment]
Note: The Latest Post block has a similar implementation for preventing opening links in the editor (ref: #40593), but it uses Snackbars instead.

Which method would be preferred for screen reader users - snackbars or description like we're adding here?

I think using one method across the blocks would set better expectations, and we could also recommend it in the future.

**Comment by alexstine** [comment]
@Mamaduka I actually like the snackbar approach because it is less verbose. Maybe this would be a good hook to create? It also makes it more clear visually with the visual notice. Something like `useDisabledNotice()`.

**Comment by Mamaduka** [comment]
@alexstine, starting with code duplication is fine for now and usually works better than early abstractions.

After we have a common pattern, I'm happy to team up and extract it into a reusable hook.

Maybe something like this:

```jsx
function Example() {
  // `useInertLinkProps` or `useDisabledLinkProps`.
  const inertLinkProps = useInertLinkProps();

  return <a className="example" { ...inertLinkProps }>Example</a>;
}
```

**Comment by alexstine** [comment]
Sounds good. I'll work on implementing this code later today. Its already after midnight. 😞

**Comment by alexstine** [comment]
@Mamaduka Finally got time to implement this. Should be good for another review.

File: packages/block-library/src/table-of-contents/list.tsx
Comment (Mamaduka) [inline packages/block-library/src/table-of-contents/list.tsx]: Do we need this? It's an internal component.

**Review by Mamaduka — COMMENTED** [review]
@alexstine, the new notice fails for the nested headings in ToC. The `TableOfContentsList` calls itself for rendering nested lists, so you'll need to pass extra props there as well.

Try testing the ToC block with the following heading structure. Clicking on "A sub heading" currently will trigger navigation.

Block code snippet:

```html
<!-- wp:heading -->
<h2 class="wp-block-heading" id="heading-text">Heading text</h2>
<!-- /wp:heading -->

<!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading" id="a-sub-heading">A sub heading</h3>
<!-- /wp:heading -->
```

File: packages/block-library/src/table-of-contents/list.tsx
Comment (alexstine) [reply packages/block-library/src/table-of-contents/list.tsx]: @Mamaduka I think so. TypeScript gets annoyed if I try to call an optional callback that might not be defined.

Comment (Mamaduka) [reply packages/block-library/src/table-of-contents/list.tsx]: Got it.

@ZebulanStanphill, do you have any suggestions?

Comment (alexstine) [reply packages/block-library/src/table-of-contents/list.tsx]: I also tried to just call `useDispatch` here first but React through an angry error because the hook was not used in the root component. Figured a consumer was the only way.

Comment (Mamaduka) [reply packages/block-library/src/table-of-contents/list.tsx]: Yeah, can’t use hooks in this component since it’s used in save method as well. No hooks are allowed there.

**Comment by alexstine** [comment]
Fixed @Mamaduka. Totally missed the last call.

File: packages/block-library/src/table-of-contents/list.tsx
Comment (ZebulanStanphill) [reply packages/block-library/src/table-of-contents/list.tsx]: Can't we just do:

```jsx
<a
onClick={ (disableLinkActivation && typeof onClick === 'function') ? onClick : undefined) }
>
```

Comment (alexstine) [reply packages/block-library/src/table-of-contents/list.tsx]: Yep, done.

**Review by Mamaduka — APPROVED** [review]
Thank you, @alexstine!

**Pull request merged**
